config/placeholder_replacement.go

```diff
@@ -181,22 +181,22 @@ func (v *placeholderReplacementVisitor) replacePackagePlaceholder(toReplace stri
 	}
 	packageType := matches[packagePlaceholderRegexp.SubexpIndex("type")]
 	packageName := matches[packagePlaceholderRegexp.SubexpIndex("name")]
-	if packageType == "" {
-		// for backwards compatibility
-		packageType = string(PackageTypeMlModel)
-	}
 	packageConfig, isPresent := v.packages[packageName]
 	if !isPresent {
 		return toReplace, errors.Errorf("failed to find a package named %q for placeholder %q",
 			packageName, toReplace)
 	}
-	if packageType != string(packageConfig.Type) {
-		expectedPlaceholder := fmt.Sprintf("packages.%s.%s", string(packageConfig.Type), packageName)
-		return toReplace,
-			errors.Errorf("placeholder %q is looking for a package of type %q but a package of type %q was found. Try %q",
-				toReplace, packageType, string(packageConfig.Type), expectedPlaceholder)
+	// package match is either of the correct type as the placeholder
+	// or (for backwards compatibility) the placeholder has no type and matches an ml_model package
+	if packageType == string(packageConfig.Type) ||
+		(packageType == "" && packageConfig.Type == PackageTypeMlModel) {
+		return packageConfig.LocalDataDirectory(viamPackagesDir), nil
 	}
-	return packageConfig.LocalDataDirectory(viamPackagesDir), nil
+
+	expectedPlaceholder := fmt.Sprintf("packages.%s.%s", string(packageConfig.Type), packageName)
+	return toReplace,
+		errors.Errorf("placeholder %q is looking for a package of type %q but a package of type %q was found. Try %q",
+			toReplace, packageType, string(packageConfig.Type), expectedPlaceholder)
```

Comment:
Right, so. I was going to explain it when I re-requested review (was still waiting for tests to pass).

I think this is the appropriate refactoring to not set a default as `PackageTypeMlModel` while also letting the `packages.%s` exists for ML models and having `packages.%s.%s` exist for all types.

Here is how it works: The variable `packageType` contains what type was parsed from the placeholder. It is allowed to be empty if there is a matching package config entry with PackageTypeMlModel, thus allow the `packages.%s` Otherwise if there is a missing type in the package config for a module, it would say:

`placeholder ${packages.module.test} is looking for a package of type '' but a package of type 'module' was found. Try '${packages..test}'`

Now that I read this, its quite confusing indeed. I think I'll add an additional check to handle when `type` is missing in the config:

if(packageConfig.Type == PackageTypeUnspecified) {
    return toReplace,
		errors.Errorf("package %q has a missing type field, please add a type to each package config.", packageConfig.name)
}

Reply:
Nevermind, I'll change to not remove the defaulting logic here. Way simpler. So no changes in this file will be done.